testcases/OpTestPCI.py

```diff
@@ -383,6 +383,131 @@ def runTest(self):
+    def runTest(self):
+        self.cv_SYSTEM.goto_state(OpSystemState.PETITBOOT_SHELL)
+        self.cv_SYSTEM.host_console_unique_prompt()
+        lspci_output = self.cv_SYSTEM.console.run_command("lspci")
+
+        # Populating device id list
+        device_ids = []
+        for line in lspci_output:
+            if line:
+                line = line.strip().split(' ')
+                device_ids.append(line[0])
+
+        class Device:
+            def __init__(self, device_info):
+                self.domain = ""
+                self.primary = ""
+                self.slotfunc = ""
+                self.secondary = ""
+                self.capability = ""
+                self.capspeed = 0
+                self.capwidth = 0
+                self.staspeed = 0
+                self.stawidth = 0
+
+                # 0000:00:00.0 PCI bridge: IBM Device 03dc
+                id_components = device_info[0].split()[0].split(":")
+                self.domain = id_components[0]
+                self.primary = id_components[1]
+                self.slotfunc = id_components[2].split()[0]
+
+                for line in device_info[1:]:
+                    if line:
+                        line = line.strip()
+                        if "Bus:" in line:
+                            line = line.split("secondary=")
+                            self.secondary = line[1][:2]
+                        if "Express (v" in line:
+                            self.capability = "Endpoint"
+                            if "Root Port" in line:
+                                self.capability = "Root"
+                            if "Upstream" in line:
+                                self.capability = "Upstream"
+                            if "Downstream" in line:
+                                self.capability = "Downstream"
+                        if "LnkCap:" in line:
+                            # LnkCap:   Port #0, Speed 8GT/s, Width x16, ASPM L0s, Exit Latency L0s unlimited, L1 unlimited
+                            line = line.split("GT/s, Width x")
+                            self.capspeed = float(line[0].split()[-1])
+                            self.capwidth = float(line[1].split(",")[0])
+                        if "LnkSta:" in line:
+                            # LnkSta:   Speed 8GT/s, Width x8, TrErr- Train- SlotClk+ DLActive+ BWMgmt- ABWMgmt+
+                            line = line.split("GT/s, Width x")
+                            self.staspeed = float(line[0].split()[-1])
+                            self.stawidth = float(line[1].split(",")[0])
+
+            def print_details(self):
+                print "%s, capability=%s, secondary=%s" %(self.get_id(), self.capability, self.secondary)
+                print "capspeed=%s, capwidth=%s, staspeed=%s, stawidth=%s" % (self.capspeed, self.capwidth, self.staspeed, self.stawidth)
+
+            def get_id(self):
+                return "%s:%s:%s" % (self.domain, self.primary, self.slotfunc)
```

Comment:
It might be worth implementing __str__ for Device and picking one of these representations, and then using `str()` to get out something human readable when printing out errors in the test case.